### PR TITLE
ci: run pyo3-ffi-check

### PR DESCRIPTION
**Pull request by messense**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -131,6 +131,32 @@ jobs:
...
+              - 'pyo3-ffi/**'
+              - '.github/workflows/ci.yml'
+              - '.github/workflows/build.yml'
+      
+      - name: Checkout pyo3-ffi-check
+        # pypy 3.7 and 3.8 are not PEP 3123 compliant so fail checks here
+        if: ${{ steps.ffi-changes.outputs.changed == 'true' && inputs.rust == 'stable' && inputs.python-version != 'pypy3.7' && inputs.python-version != 'pypy3.8' }}
+        uses: actions/checkout@v3
+        with:
+          repository: PyO3/pyo3-ffi-check
```
Comment (messense) [inline .github/workflows/build.yml:150]: Should we import the code to `pyo3` repository instead? @davidhewitt

**Comment by messense** [comment]
bors try

**Comment by messense** [comment]
bors ping

**Comment by messense** [comment]
bors try

**Comment by messense** [comment]
bors try-

**Comment by messense** [comment]
bors try

**Comment by davidhewitt** [comment]
bors try

**Comment by davidhewitt** [comment]
It seems like I got bors trying stuck with my dodgy branch 😬

File: .github/workflows/build.yml (hunk 1 above)
Comment (messense) [reply .github/workflows/build.yml]: Imported `pyo3-ffi-check` in https://github.com/PyO3/pyo3/pull/2771/commits/a14de763656735d2e623214e359727373088d583

**Comment by davidhewitt** [comment]
bors try

**Comment by davidhewitt** [comment]
bors try-

**Comment by davidhewitt** [comment]
bors try

**Comment by davidhewitt** [comment]
Oooh it's working! I deleted the `trying` branch, seems like that helped nuke bad state?!

**Comment by messense** [comment]
bors retry

**Comment by messense** [comment]
bors try

**Review by davidhewitt — APPROVED** [review]
Thanks for implementing this! I will archive the other project now.

File: pyo3-ffi-check/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+[package]
```
[the remark this answers is not in the document]
Comment (messense) [reply pyo3-ffi-check/Cargo.toml:1]: `publish = false` will do, I'll add it.

https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field

**Comment by messense** [comment]
PyPy ffi checks are changed to allow to fail for now, I'm not sure how to deal with 

> no field `ob_base` on type XXX 

errors.

**Comment by messense** [comment]
bors r=davidhewitt

**Comment by davidhewitt** [comment]
> PyPy ffi checks are changed to allow to fail for now, I'm not sure how to deal with
> `no field ob_base on type XXX`
> errors.

Yes, PyPy 3.7 and 3.8 aren't PEP 3123 compliant, they have the PyObject fields inlined, but the struct layout is the same (as far as I'm aware) and it certainly seems it works despite the fact we use `ob_base`. I decided not to worry about this for now; on Unix PyPy 3.9 is corrected, but for some reason not changed on Windows yet!

**Pull request merged**